File: emqx-mqtt-clients/scripts/emqx_list_clients.py
#!/usr/bin/env python3
"""
EMQX MQTT Agent Discovery Tool

Discover agents connected to EMQX broker. Lists all MQTT client connections
and provides agent identity, presence, and connectivity info for cross-agent communication.

Usage:
    export EMQX_HOST="localhost"
    export EMQX_API_PORT="18083"
    export EMQX_API_KEY="my-key"
    export EMQX_API_SECRET="my-secret"

    # Discover all connected agents
    python3 emqx_list_clients.py

    # Show agent connectivity endpoints
    python3 emqx_list_clients.py --endpoints

    # Watch for agent presence changes
    python3 emqx_list_clients.py --watch

    # Filter by agent name pattern
    python3 emqx_list_clients.py --search "openclaw-"

    # JSON output for programmatic use
    python3 emqx_list_clients.py --json
"""

import argparse
import json
import os
import sys
import time
import urllib.request
import urllib.error
import base64
import ssl


def get_env_or_raise(key):
    val = os.environ.get(key)
    if not val:
        print(f"Error: {key} not set. Set env var or pass via --flag.", file=sys.stderr)
        sys.exit(1)
    return val


def build_auth_header(api_key, api_secret):
    creds = f"{api_key}:{api_secret}"
    encoded = base64.b64encode(creds.encode("utf-8")).decode("utf-8")
    return f"Basic {encoded}"


def emqx_get(path, host, port, api_key, api_secret, params=None):
    """Generic EMQX API GET request."""
    base_url = f"http://{host}:{port}/api/v5{path}"
    if params:
        qs = "&".join(f"{k}={urllib.request.quote(str(v), safe='')}" for k, v in params.items())
        base_url += f"?{qs}"

    req = urllib.request.Request(
        base_url,
        headers={
            "Accept": "application/json",
            "Authorization": build_auth_header(api_key, api_secret),
        },
    )
    ctx = ssl.create_default_context()
    ctx.check_hostname = False
    ctx.verify_mode = ssl.CERT_NONE

    try:
        with urllib.request.urlopen(req, context=ctx, timeout=10) as resp:
            return json.loads(resp.read().decode("utf-8"))
    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8", errors="replace")
        print(f"HTTP {e.code}: {body}", file=sys.stderr)
        sys.exit(1)
    except urllib.error.URLError as e:
        print(f"Connection error: {e.reason} (tried: {base_url})", file=sys.stderr)
        sys.exit(1)


def fetch_clients(host, port, api_key, api_secret, limit=100, page=1, node=None, state=None, like=None):
    params = {"page": page, "limit": limit}
    if node:
        params["node"] = node
    if state:
        params["state"] = state
    if like:
        params["like"] = like
    return emqx_get("/clients", host, port, api_key, api_secret, params)


def fetch_all_clients(host, port, api_key, api_secret, limit=100, node=None, state=None, like=None):
    all_data = []
    page = 1
    meta = {}

    while True:
        resp = fetch_clients(host, port, api_key, api_secret, limit, page, node, state, like)
        data = resp.get("data", [])
        meta = resp.get("meta", {})
        all_data.extend(data)
        if not meta.get("hasnext", False):
            break
        page += 1

    return all_data, meta


def get_node_info(host, port, api_key, api_secret):
    """Get cluster node info for MQTT listener endpoints."""
    return emqx_get("/nodes", host, port, api_key, api_secret)


def build_agent_endpoints(clients, nodes):
    """Build MQTT connectivity info for each agent based on its serving node."""
    # Build node -> IP mapping from node info
    node_ips = {}
    for n in nodes:
        name = n.get("name", "")
        # Use the node's own IP, default to the connection IP
        node_ips[name] = n.get("ip", "")

    results = []
    for c in clients:
        node = c.get("node", "")
        clientid = c.get("clientid", "")
        username = c.get("username", "")
        ip = c.get("ip_address", "")
        port = c.get("port", 0)
        connected = c.get("connected", False)
        proto_ver = c.get("proto_ver", 4)
        subscriptions_cnt = c.get("subscriptions_cnt", 0)

        # Build endpoint info
        endpoint = {
            "agent_id": clientid,
            "username": username,
            "connected": connected,
            "ip": ip,
            "port": port,
            "node": node,
            "node_ip": node_ips.get(node, ""),
            "protocol": f"MQTT v{proto_ver}" if proto_ver else "MQTT",
            "subscriptions": subscriptions_cnt,
        }
        results.append(endpoint)

    return results


def format_table(agents, fields):
    """Format agent list as a readable table."""
    if not agents:
        return "No agents found."

    col_widths = {}
    for f in fields:
        key = f.get("key", f.get("label", ""))
        col_widths[key] = len(f["label"])

    for a in agents:
        for f in fields:
            key = f.get("key", f.get("label", ""))
            val = str(a.get(key, ""))
            col_widths[key] = max(col_widths[key], len(val))

    header = "  ".join(f["label"].ljust(col_widths[f.get("key", f.get("label", ""))]) for f in fields)
    sep = "  ".join("-" * col_widths[f.get("key", f.get("label", ""))] for f in fields)

    lines = [header, sep]
    for a in agents:
        row = []
        for f in fields:
            key = f.get("key", f.get("label", ""))
            val = str(a.get(key, ""))
            row.append(val.ljust(col_widths[key]))
        lines.append("  ".join(row))

    return "\n".join(lines)


def render_agent_card(agent):
    """Render a single agent's detail card."""
    lines = [
        f"┌─ Agent: {agent['agent_id']}",
        f"├ Username:      {agent['username']}",
        f"├ Status:        {'🟢 ONLINE' if agent['connected'] else '🔴 OFFLINE'}",
        f"├ MQTT Address:  {agent['ip']}:{agent['port']}",
        f"├ Node:          {agent['node']} ({agent['node_ip']})",
        f"├ Protocol:      {agent['protocol']}",
        f"└ Subscriptions: {agent['subscriptions']}",
    ]
    return "\n".join(lines)


def watch_agents(host, port, api_key, api_secret, interval=5, limit=100):
    """Watch mode: poll for agent presence changes."""
    previous = set()
    print(f"Watching EMQX agents on {host}:{port} (poll every {interval}s)...")
    print("Press Ctrl+C to stop.\n")

    try:
        while True:
            clients, meta = fetch_all_clients(host, port, api_key, api_secret, limit)
            current = {c["clientid"] for c in clients if "clientid" in c}

            joined = current - previous
            left = previous - current

            now = time.strftime("%H:%M:%S")
            for j in joined:
                client = next((c for c in clients if c["clientid"] == j), {})
                ip = client.get("ip_address", "?")
                user = client.get("username", "?")
                print(f"[{now}] ➜ JOINED  {j}  (user={user}, ip={ip})")

            for l in left:
                print(f"[{now}] ✜ LEFT    {l}")

            if not joined and not left:
                print(f"[{now}] No changes ({len(current)} agents online)")

            previous = current
            time.sleep(interval)
    except KeyboardInterrupt:
        print("\nStopped.")


def main():
    parser = argparse.ArgumentParser(
        description="EMQX MQTT Agent Discovery — find connected agents on the broker"
    )
    parser.add_argument("--host", default=None, help="EMQX host (default: $EMQX_HOST)")
    parser.add_argument("--port", default=None, help="EMQX API port (default: $EMQX_API_PORT)")
    parser.add_argument("--api-key", default=None, help="EMQX API key ID (default: $EMQX_API_KEY)")
    parser.add_argument("--api-secret", default=None, help="EMQX API secret (default: $EMQX_API_SECRET)")
    parser.add_argument("--limit", type=int, default=100, help="Page size (default: 100, max: 10000)")

    # Filters
    parser.add_argument("--node", default=None, help="Filter by EMQX node")
    parser.add_argument("--state", default=None, choices=["connected", "disconnected"],
                        help="Filter by connection state")
    parser.add_argument("--search", "--like", dest="like", default=None,
                        help="Search agent clientid by pattern")

    # Output modes
    parser.add_argument("--json", action="store_true", help="Output as JSON array")
    parser.add_argument("--raw", action="store_true", help="Output raw API response")
    parser.add_argument("--summary", action="store_true", help="Show agent count only")
    parser.add_argument("--endpoints", action="store_true",
                        help="Show with MQTT endpoint info for cross-agent communication")
    parser.add_argument("--watch", action="store_true",
                        help="Watch agent presence changes (polling)")
    parser.add_argument("--watch-interval", type=int, default=5,
                        help="Watch polling interval in seconds (default: 5)")
    parser.add_argument("--no-pager", action="store_true",
                        help="Single page only (no auto-pagination)")

    args = parser.parse_args()

    host = args.host or get_env_or_raise("EMQX_HOST")
    port = int(args.port or get_env_or_raise("EMQX_API_PORT"))
    api_key = args.api_key or get_env_or_raise("EMQX_API_KEY")
    api_secret = args.api_secret or get_env_or_raise("EMQX_API_SECRET")

    # Watch mode
    if args.watch:
        watch_agents(host, port, api_key, api_secret, args.watch_interval, args.limit)
        return

    # Fetch clients
    if args.no_pager:
        resp = fetch_clients(host, port, api_key, api_secret,
                             args.limit, 1, args.node, args.state, args.like)
        clients = resp.get("data", [])
        meta = resp.get("meta", {})
    else:
        clients, meta = fetch_all_clients(host, port, api_key, api_secret,
                                           args.limit, args.node, args.state, args.like)

    count = len(clients)
    meta_count = meta.get("count", count)

    # Enrich with node info if endpoints requested
    if args.endpoints:
        nodes = get_node_info(host, port, api_key, api_secret)
        agents = build_agent_endpoints(clients, nodes)
    else:
        agents = clients

    # Output
    if args.raw:
        print(json.dumps({"data": clients, "meta": meta}, indent=2, ensure_ascii=False))
        return

    if args.json:
        print(json.dumps(agents, indent=2, ensure_ascii=False))
        return

    if args.summary:
        label = "agents" if not args.state else f"agents ({args.state})"
        print(f"Total {label}: {count}")
        return

    if args.endpoints:
        # Agent detail cards
        for a in agents:
            print(render_agent_card(a))
            print()
        print(f"---\nTotal agents: {count}")
    else:
        # Table view - agent-focused fields
        fields = [
            {"key": "clientid", "label": "AGENT_ID"},
            {"key": "username", "label": "USERNAME"},
            {"key": "ip_address", "label": "IP"},
            {"key": "port", "label": "PORT"},
            {"key": "proto_name", "label": "PROTO"},
            {"key": "connected_at", "label": "ONLINE_SINCE"},
            {"key": "subscriptions_cnt", "label": "SUBS"},
            {"key": "node", "label": "NODE"},
        ]
        print(format_table(agents, fields))
        print(f"\nTotal agents: {count}")


if __name__ == "__main__":
    main()

File: emqx-mqtt-clients/scripts/emqx_list_clients.mjs
#!/usr/bin/env node
/**
 * EMQX MQTT Agent Discovery Tool (MQTT.js / Node.js)
 *
 * Discover agents connected to EMQX broker. Lists all MQTT client connections
 * and provides agent identity, presence, and connectivity info for cross-agent
 * communication.
 *
 * Follows mqtt-chat conventions:
 *   - Agent identity via MQTT v5 userProperties: { name, description, emoji }
 *   - Inbound topic pattern: {clientId}/inbound
 *
 * Usage:
 *   node emqx_list_clients.mjs
 *   node emqx_list_clients.mjs --endpoints
 *   node emqx_list_clients.mjs --watch
 *   node emqx_list_clients.mjs --search "openclaw-"
 *   node emqx_list_clients.mjs --json
 *   node emqx_list_clients.mjs --state connected
 *
 * Environment:
 *   EMQX_HOST          EMQX broker host
 *   EMQX_API_PORT      EMQX API port (default: 18083)
 *   EMQX_API_KEY       API key ID
 *   EMQX_API_SECRET    API key secret
 */

import { parseArgs } from "node:util";

// ── Environment helpers ────────────────────────────────────────────────

function getEnv(key) {
  return process.env[key] || "";
}

function getEnvOrRaise(key) {
  const val = getEnv(key);
  if (!val) {
    console.error(`Error: ${key} not set. Set env var or pass via --flag.`);
    process.exit(1);
  }
  return val;
}

// ── HTTP helpers ───────────────────────────────────────────────────────

function buildAuthHeader(apiKey, apiSecret) {
  const creds = Buffer.from(`${apiKey}:${apiSecret}`).toString("base64");
  return `Basic ${creds}`;
}

async function emqxGet(path, host, port, apiKey, apiSecret, params) {
  const protocol = port === 443 ? "https" : "http";
  let url = `${protocol}://${host}:${port}/api/v5${path}`;
  if (params) {
    const qs = Object.entries(params)
      .filter(([, v]) => v != null)
      .map(([k, v]) => `${k}=${encodeURIComponent(String(v))}`)
      .join("&");
    if (qs) url += `?${qs}`;
  }

  try {
    const res = await fetch(url, {
      headers: {
        Accept: "application/json",
        Authorization: buildAuthHeader(apiKey, apiSecret),
      },
      signal: AbortSignal.timeout(10000),
    });
    if (!res.ok) {
      const body = await res.text();
      console.error(`HTTP ${res.status}: ${body}`);
      process.exit(1);
    }
    return await res.json();
  } catch (err) {
    if (err.name === "TimeoutError") {
      console.error(`Connection timeout: ${url}`);
    } else {
      console.error(`Connection error: ${err.message} (tried: ${url})`);
    }
    process.exit(1);
  }
}

// ── Data fetching ──────────────────────────────────────────────────────

async function fetchClients(host, port, apiKey, apiSecret, limit = 100, page = 1, node, state, like) {
  const params = { page, limit };
  if (node) params.node = node;
  if (state) params.state = state;
  if (like) params.like = like;
  return emqxGet("/clients", host, port, apiKey, apiSecret, params);
}

async function fetchAllClients(host, port, apiKey, apiSecret, limit = 100, node, state, like) {
  const all = [];
  let page = 1;
  let meta = {};

  while (true) {
    const resp = await fetchClients(host, port, apiKey, apiSecret, limit, page, node, state, like);
    const data = resp.data || [];
    meta = resp.meta || {};
    all.push(...data);
    if (!meta.hasnext) break;
    page++;
  }

  return { clients: all, meta };
}

async function getNodeInfo(host, port, apiKey, apiSecret) {
  return emqxGet("/nodes", host, port, apiKey, apiSecret);
}

// ── Agent endpoints builder ────────────────────────────────────────────

function buildAgentEndpoints(clients, nodes) {
  const nodeIps = {};
  for (const n of nodes) {
    nodeIps[n.name || ""] = n.ip || "";
  }

  return clients.map((c) => ({
    agent_id: c.clientid || "",
    username: c.username || "",
    connected: !!c.connected,
    ip: c.ip_address || "",
    port: c.port || 0,
    node: c.node || "",
    node_ip: nodeIps[c.node || ""] || "",
    protocol: c.proto_ver ? `MQTT v${c.proto_ver}` : "MQTT",
    subscriptions: c.subscriptions_cnt || 0,
    connected_at: c.connected_at || "",
  }));
}

// ── Output formatters ──────────────────────────────────────────────────

function pad(str, len) {
  return String(str).padEnd(len);
}

function formatTable(agents, fields) {
  if (!agents.length) return "No agents found.";

  const colWidths = {};
  for (const f of fields) {
    const key = f.key || f.label;
    colWidths[key] = f.label.length;
    for (const a of agents) {
      const val = String(a[key] ?? "");
      colWidths[key] = Math.max(colWidths[key], val.length);
    }
  }

  const header = fields.map((f) => pad(f.label, colWidths[f.key || f.label])).join("  ");
  const sep = fields.map((f) => "-".repeat(colWidths[f.key || f.label])).join("  ");

  const rows = agents.map((a) =>
    fields.map((f) => pad(String(a[f.key || f.label] ?? ""), colWidths[f.key || f.label])).join("  ")
  );

  return [header, sep, ...rows].join("\n");
}

function renderAgentCard(agent) {
  const status = agent.connected ? "🟢 ONLINE" : "🔴 OFFLINE";
  return [
    `┌─ Agent: ${agent.agent_id}`,
    `├ Username:      ${agent.username}`,
    `├ Status:        ${status}`,
    `├ MQTT Address:  ${agent.ip}:${agent.port}`,
    `├ Node:          ${agent.node} (${agent.node_ip})`,
    `├ Protocol:      ${agent.protocol}`,
    `├ Connected:     ${agent.connected_at}`,
    `└ Subscriptions: ${agent.subscriptions}`,
  ].join("\n");
}

// ── Watch mode ─────────────────────────────────────────────────────────

async function watchAgents(host, port, apiKey, apiSecret, interval = 5, limit = 100) {
  let previous = new Set();
  console.log(`Watching EMQX agents on ${host}:${port} (poll every ${interval}s)...`);
  console.log("Press Ctrl+C to stop.\n");

  while (true) {
    try {
      const { clients } = await fetchAllClients(host, port, apiKey, apiSecret, limit);
      const current = new Set(clients.map((c) => c.clientid).filter(Boolean));

      const joined = [...current].filter((id) => !previous.has(id));
      const left = [...previous].filter((id) => !current.has(id));

      const now = new Date().toLocaleTimeString();
      for (const j of joined) {
        const client = clients.find((c) => c.clientid === j) || {};
        const ip = client.ip_address || "?";
        const user = client.username || "?";
        console.log(`[${now}] ➜ JOINED  ${j}  (user=${user}, ip=${ip})`);
      }
      for (const l of left) {
        console.log(`[${now}] ✜ LEFT    ${l}`);
      }
      if (!joined.length && !left.length) {
        console.log(`[${now}] No changes (${current.size} agents online)`);
      }

      previous = current;
    } catch (err) {
      console.error(`Watch error: ${err.message}`);
    }

    await new Promise((resolve) => setTimeout(resolve, interval * 1000));
  }
}

// ── Main ──────────────────────────────────────────────────────────────

async function main() {
  const { values: args } = parseArgs({
    options: {
      host:       { type: "string" },
      port:       { type: "string" },
      "api-key":  { type: "string" },
      "api-secret": { type: "string" },
      limit:      { type: "string", default: "100" },
      node:       { type: "string" },
      state:      { type: "string" },
      search:     { type: "string" },
      like:       { type: "string" },
      json:       { type: "boolean", default: false },
      raw:        { type: "boolean", default: false },
      summary:    { type: "boolean", default: false },
      endpoints:  { type: "boolean", default: false },
      watch:      { type: "boolean", default: false },
      "watch-interval": { type: "string", default: "5" },
      "no-pager": { type: "boolean", default: false },
      help:       { type: "boolean", default: false },
    },
    allowPositionals: true,
    strict: false,
  });

  if (args.help) {
    console.log(`Usage: node emqx_list_clients.mjs [options]

Options:
  --host <addr>         EMQX host (default: \$EMQX_HOST)
  --port <port>         EMQX API port (default: \$EMQX_API_PORT)
  --api-key <key>       EMQX API key ID (default: \$EMQX_API_KEY)
  --api-secret <sec>    EMQX API secret (default: \$EMQX_API_SECRET)
  --limit <n>           Page size (default: 100)
  --node <name>         Filter by node
  --state <s>           Filter by state (connected|disconnected)
  --search <pattern>    Filter by clientid pattern
  --like <pattern>      Same as --search
  --json                Output as JSON
  --raw                 Raw API response
  --summary             Agent count only
  --endpoints           Show with MQTT endpoint info
  --watch               Watch presence changes (polling)
  --watch-interval <n>  Poll interval in seconds (default: 5)
  --no-pager            Single page only`);
    return;
  }

  const host       = args.host       || getEnv("EMQX_HOST");
  const port       = parseInt(args.port || getEnv("EMQX_API_PORT") || "0", 10);
  const apiKey     = args["api-key"] || getEnv("EMQX_API_KEY");
  const apiSecret  = args["api-secret"] || getEnv("EMQX_API_SECRET");

  if (!host || !port || !apiKey || !apiSecret) {
    console.error("Error: EMQX_HOST, EMQX_API_PORT, EMQX_API_KEY, EMQX_API_SECRET must be set.");
    process.exit(1);
  }
  const limit      = parseInt(args.limit, 10);
  const node       = args.node;
  const state      = args.state;
  const like       = args.search || args.like;
  const watchInterval = parseInt(args["watch-interval"], 10);

  // Watch mode
  if (args.watch) {
    await watchAgents(host, port, apiKey, apiSecret, watchInterval, limit);
    return;
  }

  // Fetch clients
  let clients, meta;
  if (args["no-pager"]) {
    const resp = await fetchClients(host, port, apiKey, apiSecret, limit, 1, node, state, like);
    clients = resp.data || [];
    meta = resp.meta || {};
  } else {
    const result = await fetchAllClients(host, port, apiKey, apiSecret, limit, node, state, like);
    clients = result.clients;
    meta = result.meta;
  }

  const count = clients.length;
  const metaCount = meta.count || count;

  // Enrich with node info if endpoints requested
  let agents = clients;
  if (args.endpoints) {
    const nodes = await getNodeInfo(host, port, apiKey, apiSecret);
    agents = buildAgentEndpoints(clients, nodes);
  }

  // Output
  if (args.raw) {
    console.log(JSON.stringify({ data: clients, meta }, null, 2));
    return;
  }

  if (args.json) {
    console.log(JSON.stringify(agents, null, 2));
    return;
  }

  if (args.summary) {
    const label = state ? `agents (${state})` : "agents";
    console.log(`Total ${label}: ${count}`);
    return;
  }

  if (args.endpoints) {
    for (const a of agents) {
      console.log(renderAgentCard(a));
      console.log();
    }
    console.log(`---\nTotal agents: ${count}`);
  } else {
    const fields = [
      { key: "clientid",         label: "AGENT_ID" },
      { key: "username",         label: "USERNAME" },
      { key: "ip_address",       label: "IP" },
      { key: "port",             label: "PORT" },
      { key: "proto_name",       label: "PROTO" },
      { key: "connected_at",     label: "ONLINE_SINCE" },
      { key: "subscriptions_cnt", label: "SUBS" },
      { key: "node",             label: "NODE" },
    ];
    console.log(formatTable(agents, fields));
    console.log(`\nTotal agents: ${count}`);
  }
}

main().catch((err) => {
  console.error(`Fatal: ${err.message}`);
  process.exit(1);
});
